### DISCUSSION TITLE
Questionse about a makeAutoObservable implementation that handles subclasses

### DISCUSSION THREAD

**Discussion by LandonSchropp** [mobxjs/mobx#3093] [Ideas]
Currently, MobX does not support subclasses in `makeAutoObservable`. @urugator posted some of the reasons for this is #2850. 

> The situation where the extended classes are abstract (can't be instantiated) - meaning you can call makeAutoObservable only once in the "subbest" class - is kinda supportable, the problem is if the extended class is eg React.Component - in such case makeAutoObservable should ignore all it's fields and prototype.
>
> In other cases, you must be able to call makeAutoObservable multiple times on the same object - this is problematic, because we don't know which fields of the object were already annotated and which were not (this is mostly due to the ability to provide overrides option to makeAutoObservable or when you would combine it with makeObservable). It would also prevented some optimizations.

I've thought about this issue and I've been experimenting with my code to see if I can work around some of the issues. Since #2850's question is answered, I thought I'd post a new one. I hope that's okay. 🙂

In my application, I ended up writing my own implementation of `makeAutoObservable`.

``` js
function isTopLevelObjectPrototype(target) {
  return _.isNil(target) || target === Object.prototype;
}

// Recursively determines if the give property is a getter in the prototype chain of the target.
function isGetter(target, property) {
  if (isTopLevelObjectPrototype(target)) {
    return false;
  }

  return !_.isNil(Object.getOwnPropertyDescriptor(target, property)?.get)
    || isGetter(Object.getPrototypeOf(target), property);
}

// Returns true if the target property is annotated by MobX.
function isMobxAnnotatedProperty(target, property) {
  return _.isSymbol(property) && property.toString().includes("mobx")
      || isObservableProp(target, property)
      || !isGetter(target, property) && isAction(target[property]);
}
j
export function makeSubclassAutoObservable(target, prototype = target) {

  // If the target is null, we know we've hit the end of the prototype chain.
  if (isTopLevelObjectPrototype(prototype)) {
    return null;
  }

  // Grab the attributes, but exclude the constructor and any properties that are already
  // observable.
  let attributes = Reflect.ownKeys(prototype)
    |> _.reject(#, property => {
      return property === "constructor" || isMobxAnnotatedProperty(target, property);
    });

  // Convert the attribute to an object whose values are true. If a value is true, MobX
  // automatically infers the attribute's value.
  let annotations = _.fromPairs(attributes.map(attribute => [ attribute, true ]));

  // Make the attributes observable.
  makeObservable(target, annotations);

  // Recursively make the prototype observable as well.
  makeSubclassAutoObservable(target, Object.getPrototypeOf(prototype));
}
```

This can be used just like `makeAutoObservable` in the store's constructor. My implementation currently doesn't support configuration overrides, but it wouldn't be hard to add them.

```
constructor() {
  makeSubclassAutoObservable(this);
}
```

I have a few questions about this in relation to @urugator's earlier comments.

> the problem is if the extended class is eg React.Component - in such case makeAutoObservable should ignore all it's fields and prototype.

I didn't realize it was even possible to make a ReactComponent into a store using makeAutoObservable. Is that a supported use case for MobX? Is that something people do regularly? If this were a supported use case, it seems like it wouldn't be difficult to modify my above code to handle it. Am I missing something here?

> this is problematic, because we don't know which fields of the object were already annotated and which were not

In my code, I'm doing this by using `isObservableProp` and `isAction`. Is that not sufficient to determine if a property is already annotated? 

Finally, are there any other issues I'm not thinking about here?

Thanks!

**Comment by urugator** [comment]
> Is that something people do regularly

It should be possible to define *own* observables/computeds/actions basically on any class including `React.Component`. People did this regularly to define local observable component state, dunno how common it is atm.

> I didn't realize it was even possible to make a ReactComponent into a store using makeAutoObservable.

Dunno how you got to this conclusion. The problem simply is you can't take *any* class and make it automatically observable - you have to know it's internals. If you call `makeAutoObservable` in subclass, it can mess with superclass in a way, it's implementor could not anticipated.
Ideally the `makeAutoObservable` should only apply to fields that are defined in the class where it's called. Similary when you call `makeObservable` we assume that you only annotate fields that belong to that class and not inherited ones - therefore no problem here. In other words each class should annotate it's own fields.

> Is that not sufficient to determine if a property is already annotated?

No, because the parent may not call `makeAutoObservable/makeObservable` at all, or it may call `makeObservable` only for a subset of it's fields, or it may explicitely exclude a field `makeAutoObservable(this, { foo: false })`. In all these cases, the inherited field should stay non-annotated.

**Reply by LandonSchropp** [reply]
Thanks for the detailed reply!

> Dunno how you got to this conclusion.

I think I misunderstood your original comment. The use case of making observable state on class components makes sense. I'd guess that `useLocalObservable` eliminates this need for most use cases nowadays, but I can definitely see the need to still support it.

> The problem simply is you can't take any class and make it automatically observable - you have to know it's internals.

This makes sense. However, it seems to me that it'd be more helpful to call this out in the `makeAutoObservable` documentation rather than disallowing subclasses entirely. I feel like this should be the responsibility of the programmer, but when you do own the entire class hierarchy, it's very useful to be able to auto-observe it.

> Ideally the makeAutoObservable should only apply to fields that are defined in the class where it's called.

My personal use case for this is that I have a base `ModelStore` class in my application that defines shared behavior for all of my model stores. This class has about 200 lines of code and 1000 lines of tests, and I don't want to duplicate that across stores.

I can see a few different approaches to this:

1. Disallow `makeAutoObservable` on subclasses (current approach). The big downside to this is that all properties must be manually annotated, and it's easy to miss one and introduce bugs.
2. All `makeAutoObservable` on classes, but only for own properties and own prototype properties. This would neatly solve my use case.
3. Make `makeAutoObservable` recursive by default. This could cause problems as you've stated above.
4. Add a `recursive` option to `makeAutoObservable` that's `false` by default. This would make recursive behavior opt-in.

> No, because the parent may not call makeAutoObservable/makeObservable at all, or it may call makeObservable only for a subset of it's fields, or it may explicitely exclude a field makeAutoObservable(this, { foo: false }). In all these cases, the inherited field should stay non-annotated.

I think I see the design philosophy you're going for here. So you'd like each class to only worry about the properties available at its layer, and not the parent or subclass properties. In that case, would suggestion 2 above handle this?

Thanks!

**Comment by mweststrate** [comment]
For 2: If you mean 'properties declared by the class calling
makeAutoObservable' as own properties, that information doesn't exist at
runtime... Otherwise we'd definitely would have gone that route :)

For 3/4: things can get very problematic if a parent and subclass both call
makeAutoObservable. The parent's call can't see the props created by the
subclass (they don't exist yet), but can't really know for sure whether
there is a subclass going to call makeAutoObservable either.


The current design is very carefully crafted on what avoids footguns, yet
works reliably with all transpilers, environments and limited reflection
information available at runtime.

However, if the current design doesn't fit your needs, there is really
nothing that prevents building your own utility method that does roughly
the same: create a function that receives 'this', reflects on the
properties, and provides the correct config to extendObsevable(this, {
prop: decorator}). I suggest to first invest in something that solves your
specific usecase / constraints, and then standardise it if it proves to be
a generic solution to a general problem, instead of standardising it in
MobX first. Hope that helps!

**Reply by urugator** [reply]
> properties declared by the class calling
makeAutoObservable' as own properties,

It's not limited to own properties, we can't pair `makeAutoObservable` call with prototype either. 
`makeAutoObservable/makeObservable` has absolutely no clue about where it is called inside the hierarchy (in which constructor).

> own utility method

`makeObservale` is already recursive. As already noted in #2850, assuming that `ModelStore` is abstract, all that should be needed is to collect all keys from proto chain and call `makeObservable(this, { key: true, ... })` in the "subbest" class.
